util/err.c

```diff
@@ -65,3 +66,38 @@ void libutil_errx(int eval, const char *fmt, ...) {
 
     verrx(eval, fmt, args);
 }
+
+extern int getargs(void *, int, char *, int);
+
+static char buffer[PATH_MAX];
+static const char* __progname = NULL;
+
+const char* getprogname (void) {
+
+    if (!__progname) {
+        struct procentry64 entry;
+        pid_t pid = getpid();
+
+        if(getprocs64(&entry, sizeof(entry), NULL, 0, &pid, 1) > 0)
+        {
+            if(getargs(&entry, sizeof(entry), buffer, sizeof(buffer)) == 0)
+            {
+                const char *p;
+                __progname = buffer;
+                p = strrchr (__progname, '/');
+                if (p != NULL)
+                    __progname = p + 1;
+            }
+        }
+    }
+
+    return __progname;
+}
+
+void setprogname (const char* progname) {
+
+    strcpy(buffer, progname);
```

Comment:
https://www.unix.com/man-page/FreeBSD/3/getprogname/

> The setprogname() function sets the name of the program to be the last component of the progname argument.

Need to find the last `/` in the input string (eg. `strrchr`) and use the pointer from there as the program name (or the original pointer if it doesn't contain a slash).

> Since a pointer to the given string is kept as the program name, it should not be modified for the rest of the program's lifetime.

No buffer/strcpy needed. Store the pointer directly.